Update tutorial to reflect current situation

Lekensteyn opened the pull request:
Addresses almost all of the remaining comments in #2498, the only things not updated are the pictures.

Please have a closer look at the sample config, it contains a hack for a simple single-user scenario. (I do not intent to use this configuration in practice, but thought it would be a simple demonstration of auth(z) concepts).

tardyp commented on master/buildbot/scripts/sample.cfg:
```diff
@@ -68,13 +68,13 @@ c['builders'].append(
 
-# 'status' is a list of Status Targets. The results of each build will be
-# pushed to these targets. buildbot/status/*.py has a variety to choose from,
+# 'services' is a list of Reporter Targets. The status of each build will be
```
not only Reporter targets. can be any BuildbotService

tardyp commented on master/docs/manual/cfg-reporters.rst:
```diff
@@ -426,7 +426,8 @@ Patches are very welcome for restoring the full functionality.
 .. note:: Security Note
 
-Please note that any user having access to your irc channel or can PM the bot will be able to create or stop builds :bug:`3377`.
+    Please note that any user having access to your irc channel or can PM the
```
We have a one sentence per line policy, which helps for the review and the diffs please dont break sentences

tardyp commented on master/docs/tutorial/firstrun.rst:
```diff
@@ -51,27 +51,36 @@ Creating a master
 The first necessary step is to create a virtualenv for our master.
-All our operations will happen in this directory:
+We will also use a separate directory to demonstrate the separation between a
+master and worker::
```
one sentence per line

sa2ajj replied:
"separate" and "separation"?

tardyp commented on master/docs/tutorial/firstrun.rst:
```diff
@@ -51,27 +51,36 @@ Creating a master
+.. note:: *Python 3*
+
+    Python 3 support is work in progress, if you default interpreter is Python
+    3, then you must pass pass the ``--python=python3`` to ``virtualenv``.
```
--python=python2

tardyp commented on master/docs/tutorial/firstrun.rst:
```diff
@@ -51,27 +51,36 @@ Creating a master
 .. code-block:: bash
 
-  ./bin/pip install buildbot[bundle]
+  pip install buildbot[bundle]
```
while you are at it, please add a note here to update pip. We often have user having issues because of older pip

tardyp commented on master/docs/tutorial/tour.rst:
```diff
@@ -203,85 +205,61 @@ Let's tell the bot to notify certain events, to learn which EVENTS we can notify
+
+and append::
+
+  class RolesFromUsername(RolesFromBase):
```
There is a RolesFromEmails for that. I agree that RolesFromUsername is useful for minimalistic authz config, but it should be in the code (and tested), not in the doc

tardyp commented on master/docs/tutorial/tour.rst:
```diff
@@ -203,85 +205,61 @@ Let's tell the bot to notify certain events, to learn which EVENTS we can notify
+              util.AnyEndpointMatcher(role="users")
+          ],
+          rolematcher = [
+                RolesFromUsername('users', ['Alice'])
```
probably the example should say Admin instead of Alice, because that user is really the admin

sa2ajj commented on master/docs/tutorial/firstrun.rst:
```diff
@@ -51,27 +51,36 @@ Creating a master
+  cd buildbot
+  virtualenv --no-site-packages sandbox
+  source sandbox/bin/activate
+
```
one empty line too many

tardyp commented on master/buildbot/www/authz/roles.py:
```diff
@@ -70,3 +70,15 @@ def getRolesFromUser(self, userDetails, owner):
+        self.usernames = usernames
+
+    def getRolesFromUser(self, userDetails):
+        if 'username' in userDetails:
```
if userDetails.get('username') in self.usernames:

Lekensteyn replied:
In theory you can have `None` in `self.usernames` and have a match. Is that desired?

tardyp replied:
I would prefer that you ensure None not in usernames in the constructor (with a config,error)

tardyp commented on master/docs/tutorial/firstrun.rst:
```diff
@@ -51,27 +51,34 @@ Creating a master
 
-  ./bin/pip install buildbot[bundle]
+  pip install --upgrade pip
+  pip install buildbot[bundle]
```
'buildbot[bundle]' should be quoted in bash

Lekensteyn commented:
hold on, still need to make some changes.

Lekensteyn commented:
Fixed the bash issue, added the new class to setup.py and fixed the sample.cfg snippet. Tested locally and the web restriction seems to work.

Lekensteyn commented:
This might be sufficient to address/close https://github.com/buildbot/buildbot-website/issues/10 I guess? Alternatively, is a link available to the current Python 3 porting effort?

@tardyp I have replied to the `getRolesFromUser` comment.

Lekensteyn commented:
Hm, how can I use `config.error`? Should it actually be used in roles.py?

sa2ajj commented on master/docs/manual/cfg-reporters.rst:
```diff
@@ -414,6 +414,8 @@ The default ``ctx`` for the missing worker email is made of:
 
 .. index:: IRC
 
+.. _IRC-Bot:
```
should `:bb:reporter:` be used instead?

sa2ajj commented:
@Lekensteyn I think what you got here is a circular dependency: one way to resolve it is to not import `config` at the top level, but next to the place where it's used.

Lekensteyn commented:
Could not use `config.error`, presumably because it resides in `/www/`. I have changed it to use ValueError instead. Also addressed the `:bb:reporter` comment (though I did not check how the resulting docs look like).

tardyp commented:
@Lekensteyn Sorry I forgot to answer timely.

you just need to do from buildbot import config on top of the file (that is not a problem that the code is in www module)

Please don't report value error. config.error is made so that there can be several errors can be thrown and will be concatenated for a whole master.cfg diagnostic.

If you throw a value error, this will stop the master.cfg parsing immediatly.

Lekensteyn commented:
The tests fail to run when I try `from buildbot import config`, either on top of the file or in the function. I see other instances in `master/buildbot/www/` that use ValueError, but not `config.error`. Should I return to the original case where I check that the username key is set?

tardyp commented:
@Lekensteyn see https://github.com/tardyp/buildbot/commit/4a0f167cd22978b540560beae7dc7b81fca73e0a

I added a commit to use config.error. Can you update your PR?

Lekensteyn commented:
Applied a slightly different change, `config.error` does not return an exception so raising it probably did not work. Passes the local `trial buildbot/test/unit/test_www_roles.py` test.

```diff
diff --git a/master/buildbot/test/unit/test_www_roles.py b/master/buildbot/test/unit/test_www_roles.py
index a7780d8..8cb4fa4 100644
--- a/master/buildbot/test/unit/test_www_roles.py
+++ b/master/buildbot/test/unit/test_www_roles.py
@@ -13,6 +13,7 @@
 # Copyright Buildbot Team Members
 from twisted.trial import unittest
 
+from buildbot.test.util.config import ConfigErrorsMixin
 from buildbot.www.authz import roles
 
 
@@ -82,7 +83,7 @@ class RolesFromOwner(unittest.TestCase):
         self.assertEqual(ret, ["ownerofbuild"])
 
 
-class RolesFromUsername(unittest.TestCase):
+class RolesFromUsername(unittest.TestCase, ConfigErrorsMixin):
 
     def setUp(self):
         self.roles = roles.RolesFromUsername(roles=["admins"], usernames=["Admin"])
@@ -106,5 +107,5 @@ class RolesFromUsername(unittest.TestCase):
         self.assertEqual(ret, ["developers", "integrators"])
 
     def test_badUsernames(self):
-        self.assertRaises(ValueError,
-            roles.RolesFromUsername, roles=[], usernames=[None])
+        self.assertRaisesConfigError('Usernames cannot be None',
+            lambda: roles.RolesFromUsername(roles=[], usernames=[None]))
diff --git a/master/buildbot/www/authz/roles.py b/master/buildbot/www/authz/roles.py
index d50eaa5..af19912 100644
--- a/master/buildbot/www/authz/roles.py
+++ b/master/buildbot/www/authz/roles.py
@@ -76,7 +76,8 @@ class RolesFromUsername(RolesFromBase):
     def __init__(self, roles, usernames):
         self.roles = roles
         if None in usernames:
-            raise ValueError('Usernames cannot be None')
+            from buildbot import config
+            config.error('Usernames cannot be None')
         self.usernames = usernames
 
     def getRolesFromUser(self, userDetails):
```

tardyp commented:
looks good!

Lekensteyn commented:
Yay, CI is also happy now. Thank you for your assistance!

sa2ajj commented:
@Lekensteyn thank you

frozar commented:
@tardyp : Hi, do you plan to provide a new release of buildbot since this pull request updated the documentation?

tardyp commented:
@frozar I trick to stick with the 1 release per month. 
The new tour is available here anyway http://docs.buildbot.net/latest/tutorial/tour.html

frozar commented:
@tardyp : Great thank you. I'm really new to buildbot, so I follow the tutorial (I'm trying to escape from Jenkins ^^').

Frodox commented:
@Lekensteyn thank you too! :-)
